server/use_cases/add_service_commitments.py

```diff
@@ -39,56 +41,157 @@ def add_service_commitments(commitments_repo, shifts_repo, commitments):
+            if has_valid_time_attributes(
+                current_shift) and has_valid_time_attributes(
+                existing_shift):
+                if (current_shift.shift_start < existing_shift.shift_end and
+                    current_shift.shift_end > existing_shift.shift_start):
+                    has_overlap = True
+                    break
+        approved_shifts = [
+            valid_shifts[j] for j in range(i) if j in final_valid_indexes]
+        for approved_shift in approved_shifts:
+            if has_valid_time_attributes(
+                current_shift) and has_valid_time_attributes(
+                approved_shift):
+                if (current_shift.shift_start < approved_shift.shift_end and
+                    current_shift.shift_end > approved_shift.shift_start):
+                    has_overlap = True
+                    break
+        if has_overlap:
+            results[idx] = {
                 "service_commitment_id": None,
-                "success": False
+                "success": False,
+                "message": "Overlapping commitment"
             }
+        else:
+            final_valid_commitments.append(valid_commitments[i])
+            final_valid_indexes.append(idx)
+    if not final_valid_commitments:
+        for i in valid_indexes:
+            if results[i] is None:
+                results[i] = {
+                    "service_commitment_id": None,
+                    "success": False,
+                    "message": "Overlapping commitment"
+                }
         return results
-
-    # insert valid commitments
-    commitments_as_dict = [c.to_dict() for c in valid_commitments]
-    inserted_commitments = (
-        commitments_repo.insert_service_commitments(
-            (commitments_as_dict)
-        )
-    )
-
-    # fill in success results for valid commitments
-    for idx, i in enumerate(valid_indexes):
+    commitments_as_dict = [c.to_dict() for c in final_valid_commitments]
+    inserted_commitments = commitments_repo.insert_service_commitments(
+        commitments_as_dict)
+    for idx, i in enumerate(final_valid_indexes):
         if idx < len(inserted_commitments):
             results[i] = {
-                "service_commitment_id": 
-                    str(inserted_commitments[idx]),
+                "service_commitment_id": str(inserted_commitments[idx]),
                 "success": True
             }
         else:
             results[i] = {
                 "service_commitment_id": None,
-                "success": False
+                "success": False,
+                "message": "Failed to insert commitment"
+            }
+    for i in valid_indexes:
+        if results[i] is None:
+            results[i] = {
+                "service_commitment_id": None,
+                "success": False,
+                "message": "Overlapping commitment"
             }
-
     return results
 
-
 def check_time_overlap(shifts):
```

Comment:
You have a check_time_overlap function, but this function does not get used. You should use it.